File: Sorting/mergesort.c
#include <stdio.h>
#include <stdlib.h>
#define MAX 100 

int b[MAX];

void merge(int arr[], int l, int m, int r) {
  int i, j, k;
  for (i = m+1; i > l; i--)
    b[i-1] = arr[i-1];
  for (j = m; j < r; j++)
    b[r+m-j] = arr[j+1];
  for (k = l; k <= r; k++)
    arr[k] = (b[i]<b[j]) ? b[i++] : b[j--];
}

void mergesort(int arr[], int l, int r) {
  if (l < r) {
    int m = l+(r-l)/2;
    mergesort(arr, l, m);
    mergesort(arr, m+1, r);
    merge(arr, l, m, r);
  }
}

int main(int argc, const char *argv[])
{
  int a[MAX];
  int i;
  for (i = 0; i < MAX; i++) {
    a[i] = rand()%100;
    printf("%i ", a[i]);
  }
  printf("\n");

  mergesort(a,0,MAX-1);
  for (i = 0; i < MAX; i++) {
    printf("%i ", a[i]);
  }
  printf("\n");
  return 0;
}

File: Sorting/MergeSort.java
public class MergeSort {
  private static final int MAX = 200;
  private static int[] tempMergArr = new int[MAX];

  public static void merge(int arr[], int l, int m, int r) {
    for (int i = l; i <= r; i++)
      tempMergArr[i] = arr[i];
    int i = l;
    int k = l;
    int j = m+1;
    while (i <= m && j <= r)
      arr[k++] = (tempMergArr[i] <= tempMergArr[j]) ? tempMergArr[i++] : tempMergArr[j++];
    while (i <= m)
      arr[k++] = tempMergArr[i++];
  }

  public static void mergesort(int arr[], int l, int r) {
    if (l < r) {
      int m = l+(r-l)/2;
      mergesort(arr, l, m);
      mergesort(arr, m+1, r);
      merge(arr, l, m, r);
    }
  }

  public static void main (String[] args) {
    int[] array = new int[MAX];
    for (int x = 0; x != MAX; ++x) {
      array[x] = (int)(Math.random() * 100);
      System.out.print(array[x] + " ");
    } 
    System.out.println();
    mergesort(array,0,MAX-1);
    for (int x : array) {
      System.out.print(x + " ");
    } 
    System.out.println();
  }
}
